File: site/check/playground.py
#!/usr/bin/env python3
"""Drive the playground in a real browser and check it answers.

The panel is the only part of the site that *runs* rather than describes, so it
is the only part that can be broken in a way reading cannot catch: a renamed
field in the binding, a control wired to nothing, a wasm bundle that never
loads. The quickstart checker executes the page's code snippets for the same
reason; this executes the page.

    python3 site/check/playground.py

Needs the wasm built (`sh site/build-wasm.sh`) and Playwright's Chromium,
which the demo's frontend already depends on.
"""

from __future__ import annotations

import http.server
import json
import socket
import subprocess
import sys
import threading
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent.parent
SITE = ROOT / "site"
#: The demo's frontend is where Playwright lives; running from there keeps a
#: second copy of a heavyweight dependency out of the tree.
RUNNER = ROOT / "examples" / "explorer" / "web"

DRIVER = """
import { chromium } from "playwright";
const url = process.argv[2];
const browser = await chromium.launch();
const page = await browser.newPage();
const problems = [];
page.on("pageerror", (e) => problems.push(`page error: ${e}`));
page.on("console", (m) => { if (m.type() === "error") problems.push(`console: ${m.text()}`); });

// Record every request, so "the bundle is not fetched until asked for" is
// an observation rather than a claim about the code.
const fetched = [];
page.on("request", (r) => fetched.push(r.url()));

await page.goto(url, { waitUntil: "load" });

// Deliberately *before* scrolling or clicking: at this point the panel is
// below the fold and the wasm must not have been requested.
await page.waitForTimeout(400);
const eagerlyFetched = fetched.filter((u) => u.endsWith(".wasm"));

// Clicked only if it is still clickable. If something already started the
// load the button is disabled by now, and clicking would time out with a
// message about element stability rather than about the thing that is
// actually wrong — which is what the `eagerlyFetched` count below reports.
const load = page.locator('[data-play="load"]');
if (await load.isEnabled()) await load.click();

// The panel stays hidden until the module loads, so its visibility *is* the
// "did wasm come up" assertion.
await page.waitForSelector(".play:not([hidden])", { timeout: 60000 });
const lazilyFetched = fetched.filter((u) => u.endsWith(".wasm"));

const read = async () => {
  await page.waitForTimeout(120);
  return {
    badges: await page.locator('[data-play="badges"] .badge').allInnerTexts(),
    plan: await page.locator('[data-play="plan"]').innerText(),
    status: await page.locator('[data-play="status"]').innerText(),
    rows: await page.locator('[data-play="rows"] tbody tr').count(),
    columns: await page.locator('[data-play="columns"] input').count(),
  };
};

// 1. It loads and answers something.
const initial = await read();

// 2. Filtering on the indexed column, reading every column: a table scan,
//    because the point reads an index scan implies cost more than the scan.
await page.locator('[data-play="column"]').selectOption({ index: 1 });
await page.locator('[data-play="op"]').selectOption("eq");
await page.locator('[data-play="value"]').fill("1");
const scan = await read();

// 3. Narrow to the indexed column alone and the plan becomes index-only.
const boxes = page.locator('[data-play="columns"] input');
const count = await boxes.count();
for (let i = 0; i < count; i++) if (i !== 1) await boxes.nth(i).uncheck();
const covering = await read();

// 4. A refusal is shown, not swallowed.
for (let i = 0; i < count; i++) await boxes.nth(i).check();
await page.locator('[data-play="value"]').fill("not-a-number");
const refused = await read();

// 5. A write, and the index answering for it. The panel's whole claim about
//    being a record layer rather than a query engine.
await page.locator('[data-play="value"]').fill("2");
const beforeWrite = await read();
const fields = page.locator('[data-play="fields"] input');
for (const [i, v] of ["9100", "2", "A New Book", "2001"].entries()) {
  await fields.nth(i).fill(v);
}
await page.locator('[data-play="insert"]').click();
const afterWrite = await read();

// 6. Two conditions.
await page.locator('[data-play="value"]').fill("1");
await page.locator('[data-play="column2"]').selectOption({ index: 3 });
await page.locator('[data-play="op2"]').selectOption("gt");
await page.locator('[data-play="value2"]').fill("1970");
const conjunction = await read();

// 7. The join panel.
await page.locator(".play-join summary").click();
await page.locator('[data-play="groupby"]').selectOption("2");
await page.locator('[data-play="runjoin"]').click();
await page.waitForTimeout(300);
const join = {
  plan: await page.locator('[data-play="joinplan"]').innerText(),
  status: await page.locator('[data-play="joinstatus"]').innerText(),
  rows: await page.locator('[data-play="joinrows"] tbody tr').count(),
};

console.log(JSON.stringify({
  beforeWrite, afterWrite, conjunction, join,
  initial, scan, covering, refused, problems,
  eagerlyFetched: eagerlyFetched.length,
  lazilyFetched: lazilyFetched.length,
}));
await browser.close();
"""


def free_port() -> int:
    with socket.socket() as s:
        s.bind(("127.0.0.1", 0))
        return int(s.getsockname()[1])


def main() -> int:
    for needed in ("slate_wasm.js", "slate_wasm_bg.wasm"):
        if not (SITE / needed).exists():
            print(f"FAIL  {needed} is not built; run `sh site/build-wasm.sh`")
            return 1

    port = free_port()
    handler = lambda *a, **k: http.server.SimpleHTTPRequestHandler(  # noqa: E731
        *a, directory=str(SITE), **k
    )
    server = http.server.ThreadingHTTPServer(("127.0.0.1", port), handler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()

    driver = RUNNER / "playground-check.mjs"
    driver.write_text(DRIVER)
    try:
        result = subprocess.run(
            ["node", str(driver), f"http://127.0.0.1:{port}/index.html"],
            cwd=RUNNER,
            capture_output=True,
            text=True,
            timeout=180,
        )
    finally:
        driver.unlink(missing_ok=True)
        server.shutdown()

    if result.returncode != 0:
        print("FAIL  the browser driver exited non-zero")
        print(result.stdout.strip())
        print(result.stderr.strip())
        return 1

    seen = json.loads(result.stdout.strip().splitlines()[-1])
    failures: list[str] = []

    def check(name: str, ok: bool, detail: str) -> None:
        print(f"{'ok  ' if ok else 'FAIL'}  {name}")
        if not ok:
            failures.append(name)
            print(f"        {detail}")

    check(
        "the wasm module loads and the panel answers",
        seen["initial"]["rows"] > 0 and bool(seen["initial"]["plan"]),
        f"initial: {seen['initial']}",
    )
    check(
        "a filter on the indexed column plans as a table scan",
        "Table Scan" in seen["scan"]["plan"],
        f"got {seen['scan']['plan']!r}",
    )
    check(
        "narrowing to the indexed column alone makes it index-only",
        "Index Only Scan" in seen["covering"]["plan"],
        f"got {seen['covering']['plan']!r}",
    )
    check(
        "and the covering plan is marked as the good one",
        any("Index Only" in b for b in seen["covering"]["badges"]),
        f"badges: {seen['covering']['badges']}",
    )
    check(
        "the plan's decoded columns narrow with the projection",
        any("decodes [1]" in b for b in seen["covering"]["badges"]),
        f"badges: {seen['covering']['badges']}",
    )
    check(
        "a bad literal is refused in the panel rather than swallowed",
        "whole number" in seen["refused"]["status"],
        f"status: {seen['refused']['status']!r}",
    )
    check(
        "the wasm is not fetched until the reader asks for it",
        seen["eagerlyFetched"] == 0,
        f"{seen['eagerlyFetched']} wasm request(s) before the button was clicked",
    )
    check(
        "and it is fetched once they do",
        seen["lazilyFetched"] >= 1,
        "the panel came up without fetching any wasm, which cannot be right",
    )
    check(
        "an inserted row appears in the current query",
        seen["afterWrite"]["rows"] == seen["beforeWrite"]["rows"] + 1,
        f"{seen['beforeWrite']['rows']} then {seen['afterWrite']['rows']}",
    )
    check(
        "a second condition narrows the result",
        seen["conjunction"]["rows"] == 3,
        f"Le Guin after 1970 should be three books, got {seen['conjunction']['rows']}",
    )
    check(
        "the join panel groups and shows a join plan",
        "Join" in seen["join"]["plan"] and seen["join"]["rows"] >= 1,
        f"{seen['join']}",
    )
    check("no page or console errors", not seen["problems"], f"{seen['problems']}")

    print()
    if failures:
        print(f"{len(failures)} failed: {', '.join(failures)}")
        return 1
    print("the playground runs the kernel in a browser")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: site/playground.js
// The playground: the record layer's kernel, in this tab.
//
// `slate_wasm.js` and `slate_wasm_bg.wasm` are built from `crates/slate-wasm`
// and are *not* committed — CI builds them and the Pages deploy carries them,
// so a stale binary cannot drift from the kernel it claims to be. The panel
// stays hidden until the module loads, so a reader who arrives before the
// build lands (or with wasm disabled) sees the prose and not a dead widget.

const panel = document.querySelector("[data-playground]");
if (panel) {
  const el = (name) => panel.querySelector(`[data-play="${name}"]`);

  const state = { schema: [], playground: null, columns: new Set() };

  const show = (message) => {
    el("status").textContent = message;
  };

  /** Rebuild the column and sort pickers for the selected table. */
  function describeTable() {
    const table = state.schema.find((t) => t.name === el("table").value);
    if (!table) return;

    for (const select of [el("column"), el("sort")]) {
      select.innerHTML = "";
      if (select === el("sort")) {
        select.append(new Option("(none)", ""));
      }
      for (const column of table.columns) {
        // Indexed columns are marked, because which one you pick is the
        // whole point of the exercise.
        const mark = table.indexed.includes(column.ordinal) ? " ·idx" : "";
        select.append(new Option(`${column.name}${mark}`, String(column.ordinal)));
      }
    }

    // Every column selected by default: that is the shape that *cannot* be
    // index-only, so the reader starts from the uninteresting plan and
    // narrows towards the interesting one.
    state.columns = new Set(table.columns.map((c) => c.ordinal));
    const boxes = el("columns");
    boxes.innerHTML = "";
    for (const column of table.columns) {
      const label = document.createElement("label");
      const box = document.createElement("input");
      box.type = "checkbox";
      box.checked = true;
      box.addEventListener("change", () => {
        if (box.checked) state.columns.add(column.ordinal);
        else state.columns.delete(column.ordinal);
        run();
      });
      label.append(box, document.createTextNode(column.name));
      boxes.append(label);
    }
  }

  function spec() {
    const table = state.schema.find((t) => t.name === el("table").value);
    const all = table ? table.columns.length : 0;
    const value = el("value").value;
    const limit = Number.parseInt(el("limit").value, 10);
    const sort = el("sort").value;

    return {
      table: el("table").value,
      // An empty value means no filter at all, rather than a filter against
      // the empty string — which would quietly return nothing and look like
      // a bug in the database.
      ...(value === "" ? {} : { filter: { column: Number(el("column").value), op: el("op").value, value } }),
      ...(sort === "" ? {} : { sort: [{ column: Number(sort), descending: el("direction").value === "desc" }] }),
      ...(Number.isFinite(limit) && limit > 0 ? { limit } : {}),
      // Sending every column is the same as sending none, and the kernel
      // reads "no projection" as "all of them" — so send [] and let it say so.
      columns: state.columns.size === all ? [] : [...state.columns].sort((a, b) => a - b),
    };
  }

  function run() {
    if (!state.playground) return;
    const answer = JSON.parse(state.playground.run(JSON.stringify(spec())));

    if (answer.error) {
      el("badges").innerHTML = "";
      el("plan").textContent = "";
      el("rows").innerHTML = "";
      show(answer.error);
      return;
    }

    const plan = answer.plan;
    const badges = [
      [plan.access, plan.indexOnly ? "good" : null],
      [`decodes [${plan.decodes.join(", ")}]`, null],
      [`cost ${plan.estimatedCost.toFixed(3)}`, null],
      [`est. ${Math.round(plan.estimatedRows)} rows`, null],
      plan.sorts ? ["sorts", "warn"] : null,
    ].filter(Boolean);

    el("badges").innerHTML = "";
    for (const [text, tone] of badges) {
      const span = document.createElement("span");
      span.className = "badge";
      if (tone) span.dataset.tone = tone;
      span.textContent = text;
      el("badges").append(span);
    }
    el("plan").textContent = plan.display;

    const table = state.schema.find((t) => t.name === el("table").value);
    const shown = spec().columns;
    const headers = shown.length
      ? shown.map((o) => table.columns[o].name)
      : table.columns.map((c) => c.name);

    // Rows always come back whole; a projection changes what the *plan*
    // decodes, not the shape of the answer. Showing only the projected
    // columns would hide that, so the table shows everything and the
    // `decodes` badge above is where the narrowing is visible.
    const head = `<tr>${table.columns.map((c) => `<th>${c.name}</th>`).join("")}</tr>`;
    const body = answer.rows
      .map((row) => `<tr>${row.map((v) => `<td>${escape(v)}</td>`).join("")}</tr>`)
      .join("");
    el("rows").innerHTML = `<table><thead>${head}</thead><tbody>${body}</tbody></table>`;

    const narrowed = headers.length < table.columns.length ? `, reading ${headers.join(", ")}` : "";
    show(`${answer.returned} row${answer.returned === 1 ? "" : "s"}${narrowed}`);
  }

  const escape = (text) =>
    String(text).replace(/[&<>"]/g, (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;" })[c]);

  (async () => {
    try {
      const module = await import("./slate_wasm.js");
      await module.default();
      state.playground = new module.Playground();
      state.schema = JSON.parse(state.playground.schema());

      const tables = el("table");
      for (const table of state.schema) tables.append(new Option(table.name, table.name));
      tables.value = "books";

      describeTable();
      tables.addEventListener("change", () => {
        el("value").value = "";
        describeTable();
        run();
      });
      for (const name of ["column", "op", "value", "sort", "direction", "limit"]) {
        const control = el(name);
        control.addEventListener(control.tagName === "SELECT" ? "change" : "input", run);
      }
      el("all").addEventListener("click", () => {
        describeTable();
        run();
      });

      panel.hidden = false;
      run();
    } catch (error) {
      // Left hidden rather than showing a broken panel: the prose above still
      // explains what the playground would have shown.
      console.error("the playground did not load", error);
    }
  })();
}
